packages/react-devtools-extensions/src/background.js

```diff
@@ -78,17 +78,40 @@ function setIconAndPopup(reactBuildType, tabId) {
   });
 }
 
-// Listen to URL changes on the active tab and reset the DevTools icon.
-// This prevents non-disabled icons from sticking in Firefox.
-// Don't listen to this event in Chrome though.
-// It fires more frequently, often after onMessage() has been called.
-if (IS_FIREFOX) {
-  chrome.tabs.onUpdated.addListener((tabId, changeInfo, tab) => {
+function isRestrictedBrowserPage(url) {
+  return !url || new URL(url).protocol === 'chrome:';
+}
+
+function checkAndHandleRestrictedPageIfSo(tab) {
+  if (tab && isRestrictedBrowserPage(tab.url)) {
+    setIconAndPopup('restricted', tab.id);
+  }
+}
+
+// update popup page of any existing open tabs, if they are restricted browser pages.
+// we can't update for any other types (prod,dev,outdated etc)
+// as the content script needs to be injected at document_start itself for those kinds of detection
+// TODO: Show a different popup page(to reload current page probably) for old tabs, opened before the extension is installed
+if (!IS_FIREFOX) {
```

Comment:
Why not Firefox? Seems like it works okay with a few tweaks:

<img width="334" alt="Screen Shot 2020-10-14 at 12 05 40 PM" src="https://user-images.githubusercontent.com/29597/96015493-9e6d5700-0e15-11eb-9b80-40cbf0c5b7ba.png">

```diff
diff --git a/packages/react-devtools-extensions/src/background.js b/packages/react-devtools-extensions/src/background.js
index 547221a6bb..9fa9ba3428 100644
--- a/packages/react-devtools-extensions/src/background.js
+++ b/packages/react-devtools-extensions/src/background.js
@@ -79,7 +79,18 @@ function setIconAndPopup(reactBuildType, tabId) {
 }
 
 function isRestrictedBrowserPage(url) {
-  return !url || new URL(url).protocol === 'chrome:';
+  if (!url) {
+    return true;
+  }
+
+  const protocol = new URL(url).protocol;
+  switch (protocol) {
+    case 'chrome:':
+    case 'about:':
+      return true;
+  }
+
+  return false;
 }
 
 function checkAndHandleRestrictedPageIfSo(tab) {
@@ -92,12 +103,10 @@ function checkAndHandleRestrictedPageIfSo(tab) {
 // we can't update for any other types (prod,dev,outdated etc)
 // as the content script needs to be injected at document_start itself for those kinds of detection
 // TODO: Show a different popup page(to reload current page probably) for old tabs, opened before the extension is installed
-if (!IS_FIREFOX) {
-  chrome.tabs.query({}, tabs => tabs.forEach(checkAndHandleRestrictedPageIfSo));
-  chrome.tabs.onCreated.addListener((tabId, changeInfo, tab) =>
-    checkAndHandleRestrictedPageIfSo(tab),
-  );
-}
+chrome.tabs.query({}, tabs => tabs.forEach(checkAndHandleRestrictedPageIfSo));
+chrome.tabs.onCreated.addListener((tabId, changeInfo, tab) =>
+  checkAndHandleRestrictedPageIfSo(tab),
+);
 
 chrome.tabs.onUpdated.addListener((tabId, changeInfo, tab) => {
   // Listen to URL changes on the active tab and reset the DevTools icon.
@@ -108,9 +117,8 @@ chrome.tabs.onUpdated.addListener((tabId, changeInfo, tab) => {
     if (tab.active && changeInfo.status === 'loading') {
       setIconAndPopup('disabled', tabId);
     }
-  } else {
-    checkAndHandleRestrictedPageIfSo(tab);
   }
+  checkAndHandleRestrictedPageIfSo(tab);
 });
 
 chrome.runtime.onMessage.addListener((request, sender) => {
```

Reply:
well so basically in firefox, I am not getting the url property in the tab object, when doing chrome tabs query. so have skipped for now
Also firefox special pages are loaded in about: protocols, like about:preferences thus making it more tricky to detect restricted pages
Maybe can work on firefox support in next PR, but as chrome has huge (2 million+) users, so wanted to do that only

Reply:
I'm not sure who suggested making "edge:" a protocol we check for. Just "about:" and "chrome:" which seem pretty standard.

Reply:
Seems like the problem is that in Firefox, we don't get an update/event when switching from e.g. `about://addons` to another URL?

Reply:
err no, it is that we don't get the url at all when doing something like,
`chrome.tabs.query({}, (tabs=>tabs.forEach(tab=>console.log(tab.url))))`
in firefox, all we get are undefined console logs. Maybe you can try on your machine and tell me output ?

Reply:
https://github.com/facebook/react/pull/20023#discussion_r504823497 well that's fine, but there may be a chrome:// url in firefox which is not a restricted page right?

Reply:
My question isn't so much about the specific API (e.g. `chrome.tabs.query`) as what the observable behavior is. I believe on Firefox the observable behavior would be fine, except for the case of going from a protected tab to an unprotected one. I think in this case, it may leave the icon stale.

Reply:
sorry I am not able to understand you fully, so on firefox unfortunately what I have seen is if we do the restricted page check code for firefox, it ends up marking every page as a restricted/protected page. definitely not what we want

Reply:
> sorry I am not able to understand you fully

It's okay 😄 

> it ends up marking every page as a restricted/protected page

Interesting. That does not match what I saw when testing, but it did break after navigating from `about:addons` to `reactjs.org` so it doesn't work fully regardless. Bummer.

Reply:
😊 . so on navigating to about:addons to reactjs.org , we are not getting the latest url is it ? what about in the `changeInfo` object , nothing there also ?

Reply:
Maybe I misinterpreted what I was seeing when testing. Seems like there's some timing issue. Sometimes it seems to work okay in Firefox, but most of the time there's a false positive (the extension thinking it's a restricted page when it's not).

Reply:
I see, maybe the timing issue is that the page on unload sends a clear icon message and also the onupdated event gets triggered,but they don't trigger in proper order

Reply:
Maybe we can do something like `getPopup` to see what popup is set for the particular tabId to solve this then ?